Added 'applyGradient' method to UIViewExtensions

JayMehta97 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
🚀
## Checklist
<!--- Please go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 5.0.
- [x] New extensions support iOS 8.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+, or use `@available` if not.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

LucianoPAlmeida commented:
@guykogus  +1 to this option

JayMehta97 commented:
> Actually this seems to me like it would be better off as a `CALayerGradient` extension for initialising all the values. I.e.
> 
> ```swift
> extension CAGradientLayer {
> 
>     convenience init(colors: [Color], locations: [CGFloat]? = nil, startPoint: CGPoint, endPoint: CGPoint, type: CAGradientLayerType = .axial) {
>         self.init()
>         self.colors =  colors.map { $0.cgColor }
>         self.locations = locations?.map { NSNumber(value: Double($0)) }
>         self.startPoint = startPoint
>         self.endPoint = endPoint
>         self.type = type
>     }
> 
> }
> ```

I agree a `CAGradientLayer` extension seems better but this makes a two step process for user to have gradient background. First user has to create a `CAGradientLayer` with this init and then insert it into the `layer` as sublayer. And I also wanted developers who are unaware of the `CAGradientLayer` can simply call the method and have gradient background.

guykogus commented:
I think it's OK for this to be separated into 2 steps. If you want a method to show people how to apply a gradient, you can create a simpler version of your function:
```Swift
func applyGradient(_ gradientLayer: CAGradientLayer) {
    gradientLayer.frame = bounds
    layer.insertSublayer(gradientLayer, at: 0)
}
```
This would make the usage still fairly simple.
```Swift
view.applyGradient(CAGradientLayer(colors: [...], ...))
```

JayMehta97 commented:
@guykogus @chulbert @LucianoPAlmeida 

Thanks for your reviews. I will make these changes.

Under which group should I create CAGradientLayer extension?

guykogus commented:
@JayMehta97 you'll actually need to create a new group for `CoreAnimation`. This will include a new `Sources` folder, `Tests` folder as well as updates to the podspec and documentation.

JayMehta97 commented:
I have made and pushed suggested changes.

Changes:-
1. Created new CoreAnimation group and wrote a convenience init in CAGradientLayerExtensions.swift.
2. Created new CoreAnimationTests group and wrote a convenience init XCTest in CAGradientLayerExtensionsTests.swift. 
3. Updated CHANGELOG.md
4. Updated README.md
5. Updated pod spec file.

JayMehta97 commented:
@LucianoPAlmeida @guykogus 

Should I open another PR?

guykogus commented on CHANGELOG.md:
```diff
@@ -13,6 +13,8 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
   - Added `enumerateMatches(in:options:range:using:)`, `matches(in:options:range:)`, `numberOfMatches(in:options:range:)`, `firstMatch(in:options:range:)`, `rangeOfFirstMatch(in:options:range:)`, `stringByReplacingMatches(in:options:range:withTemplate:)`, `replaceMatches(in:options:range:withTemplate:)`, which use `String` and `String.Range` in place of `NSString` and `NSRange` to make the calls  ...
 - **UIImage**:
   - Added `withBackgroundColor(_:)` to specify a background color for a partially transparent image. [#721](https://github.com/SwifterSwift/SwifterSwift/pull/721) by [MaxHaertwig](https://github.com/maxhaertwig).
+  - **CAGradientLayer**:
```
```suggestion
- **CAGradientLayer**:
```

guykogus commented on CHANGELOG.md:
```diff
@@ -13,6 +13,8 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
 - **UIImage**:
   - Added `withBackgroundColor(_:)` to specify a background color for a partially transparent image. [#721](https://github.com/SwifterSwift/SwifterSwift/pull/721) by [MaxHaertwig](https://github.com/maxhaertwig).
+  - **CAGradientLayer**:
+    - Added `init(colors:locations:startPoint:endPoint:type:)` convenience initializers. [#726](https://github.com/SwifterSwift/SwifterSwift/pull/726) by [JayMehta97](https://github.com/JayMehta97).
```
```suggestion
  - Added `init(colors:locations:startPoint:endPoint:type:)` convenience initializer. [#726](https://github.com/SwifterSwift/SwifterSwift/pull/726) by [JayMehta97](https://github.com/JayMehta97).
```

guykogus commented on Sources/SwifterSwift/CoreAnimation/CAGradientLayerExtensions.swift:
```diff
@@ -0,0 +1,35 @@
+    ///                interpolated. (default is nil)
+    /// - Parameter startPoint: start point corresponds to the first gradient stop (I.e. [0,0] is the bottom-corner of the layer, [1,1] is the                          top-right corner.)
+    /// - Parameter endPoint: end point corresponds to the last gradient stop
+    /// - Parameter type: The kind of gradient that will be drawn. Currently, the only allowed values are `axial' (the default value),                            `radial', and `conic'.
```
```suggestion
    /// - Parameter type: The kind of gradient that will be drawn. Currently, the only allowed values are `axial' (the default value), `radial', and `conic'.
```

guykogus commented on Sources/SwifterSwift/CoreAnimation/CAGradientLayerExtensions.swift:
```diff
@@ -0,0 +1,35 @@
+    ///                assumed to spread uniformly across the [0,1] range. When rendered,
+    ///                the colors are mapped to the output colorspace before being
+    ///                interpolated. (default is nil)
+    /// - Parameter startPoint: start point corresponds to the first gradient stop (I.e. [0,0] is the bottom-corner of the layer, [1,1] is the                          top-right corner.)
```
```suggestion
    /// - Parameter startPoint: start point corresponds to the first gradient stop (I.e. [0,0] is the bottom-corner of the layer, [1,1] is the top-right corner.)
```

guykogus commented on Tests/CoreAnimationTests/CAGradientLayerExtensionsTests.swift:
```diff
@@ -0,0 +1,25 @@
+
+    func testInitWithGradientAttributes() {
+        let gradientLayer = CAGradientLayer(colors: [.red, .blue, .orange, .yellow], locations: nil, startPoint: CGPoint(x: 0.0, y: 0.5), endPoint: CGPoint(x: 1.0, y: 0.5), type: .axial)
+
```
Please add tests to check the rest of the values

JayMehta97 replied:
Sure

guykogus commented:
@JayMehta97 I think it's best to wait for this PR to be merged before creating the new one

JayMehta97 commented:
@guykogus 

I have made all the suggested changes.

LucianoPAlmeida commented:
@JayMehta97  Lixux build is failing, in this case just as if !os(Linux) should be good :)

JayMehta97 commented:
@LucianoPAlmeida 

Added the Linux check.

guykogus commented on Tests/CoreAnimationTests/CAGradientLayerExtensionsTests.swift:
```diff
@@ -0,0 +1,34 @@
+
+    func testInitWithGradientAttributes() {
+        let colors: [UIColor] = [.red, .blue, .orange, .yellow]
+        let locations: [CGFloat]? = nil
```
Test a non-nil value

guykogus commented on Tests/CoreAnimationTests/CAGradientLayerExtensionsTests.swift:
```diff
@@ -0,0 +1,34 @@
+        let locations: [CGFloat]? = nil
+        let startPoint = CGPoint(x: 0.0, y: 0.5)
+        let endPoint = CGPoint(x: 1.0, y: 0.5)
+        let gradientLayerType: CAGradientLayerType = .axial
```
```suggestion
        let gradientLayerType = CAGradientLayerType.axial
```

guykogus commented on Tests/CoreAnimationTests/CAGradientLayerExtensionsTests.swift:
```diff
@@ -0,0 +1,34 @@
+        let gradientLayer = CAGradientLayer(colors: colors, locations: locations, startPoint: startPoint, endPoint: endPoint, type: gradientLayerType)
+
+        XCTAssertEqual(gradientLayer.colors?.count, colors.count)
+        XCTAssertEqual(gradientLayer.locations as? [CGFloat], locations)
```
This works for now because `locations` is `nil`, but you'll need to `map` the values

JayMehta97 replied:
I had tried it with non-nil values and it worked fine.

SD10 commented:
Thank you for contributing to SwifterSwift! I've invited you to join the SwifterSwift GitHub organization - no pressure to accept! If you'd like more information on what that means, check out our [contributing guidelines](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md). Feel free to reach out if you have any questions! 😃

JayMehta97 commented:
@SD10 , Thanks for the invitation. Can you please brief me more about what it means.

@LucianoPAlmeida , Thanks for merging my PR. It is my first contribution to any open source repo. Looking forward to contribute more to this.

LucianoPAlmeida commented:
> @SD10 , Thanks for the invitation. Can you please brief me more about what it means.

@JayMehta97 That's an automatic message, where for every contributor when it gets its first PR merged, they are added to the organization. 
This basically means that you now is a member o SwifterSwift/contributors group, and you have some rights on the repo e.g. you can create branches on origin, instead of your fork... but most important is you are now part of the organization \o/
